sqld/src/hrana/session.rs

```diff
@@ -0,0 +1,193 @@
+    let mut stmt_iter = Statement::parse(&proto_stmt.sql);
+    let stmt = match stmt_iter.next() {
+        Some(stmt_res) => stmt_res?,
+        None => bail!(ResponseError::SqlNoStmt),
+    };
+
+    if stmt_iter.next().is_some() {
+        bail!(ResponseError::SqlManyStmts)
+    }
+
+    let params = proto_stmt
+        .args
+        .into_iter()
+        .map(proto_value_to_value)
+        .collect();
+    let params = Params::Positional(params);
+    Ok(Query { stmt, params })
+}
+
+fn proto_stmt_result_from_query_response(query_response: QueryResponse) -> proto::StmtResult {
+    let QueryResponse::ResultSet(result_set) = query_response;
+    let proto_cols = result_set
+        .columns
+        .into_iter()
+        .map(|col| proto::Col {
+            name: Some(col.name),
+        })
+        .collect();
+    let proto_rows = result_set
+        .rows
+        .into_iter()
+        .map(|row| row.values.into_iter().map(proto_value_from_value).collect())
+        .collect();
+    proto::StmtResult {
+        cols: proto_cols,
+        rows: proto_rows,
+        affected_row_count: result_set.affected_row_count,
+    }
+}
+
+fn proto_value_to_value(proto_value: proto::Value) -> Value {
+    match proto_value {
+        proto::Value::Null => Value::Null,
+        proto::Value::Integer { value } => Value::Integer(value),
+        proto::Value::Float { value } => Value::Real(value),
+        proto::Value::Text { value } => Value::Text(value),
+        proto::Value::Blob { value } => Value::Blob(value),
+    }
+}
+
+fn proto_value_from_value(value: Value) -> proto::Value {
+    match value {
+        Value::Null => proto::Value::Null,
+        Value::Integer(value) => proto::Value::Integer { value },
+        Value::Real(value) => proto::Value::Float { value },
+        Value::Text(value) => proto::Value::Text { value },
+        Value::Blob(value) => proto::Value::Blob { value },
+    }
+}
+
+fn proto_response_error_from_error(error: Error) -> Result<ResponseError, Error> {
+    Ok(match error {
+        Error::LibSqlInvalidQueryParams(source) => ResponseError::InvalidArgs { source },
+        Error::LibSqlTxTimeout(_) => ResponseError::TransactionTimeout,
+        Error::LibSqlTxBusy => ResponseError::TransactionBusy,
+        Error::RusqliteError(rusqlite_error) => match rusqlite_error {
+            rusqlite::Error::SqliteFailure(sqlite_error, message) => ResponseError::SqliteError {
+                source: sqlite_error,
+                message,
+            },
+            rusqlite::Error::SqlInputError {
+                error: sqlite_error,
+                msg: message,
+                offset,
+                ..
+            } => ResponseError::SqlInputError {
+                source: sqlite_error,
+                message,
+                offset,
+            },
+            rusqlite_error => return Err(Error::RusqliteError(rusqlite_error)),
+        },
+        error => return Err(error),
```

Comment:
same here

Reply:
Ok. Though I have to say, I didn't get this signature right on the first try :)
```rust
impl TryFrom<Error> for ResponseError {
    type Error = Error;
    fn try_from(error: Error) -> Result<ResponseError, Error>;
}
```